multiqc/modules/biscuit/biscuit.py

```diff
@@ -0,0 +1,688 @@
+#!/usr/bin/env python
+
+""" MultiQC module to parse output from BISCUIT """
+
+from __future__ import print_function
+from collections import OrderedDict
+import logging
+import re
+
+from multiqc import config
+from multiqc.plots import beeswarm, linegraph, bargraph, table
+from multiqc.modules.base_module import BaseMultiqcModule
+
+# Initialise the logger
+log = logging.getLogger(__name__)
+
+def _uniquify_ID(data, header, namespace):
+    header2 = dict([(namespace+k, v) for k, v in header.items()])
+    data2 = {}
+    for sid, datum in data.items():
+        data2[sid] = dict([(namespace+k,v) for k, v in datum.items()])
+
+    return (data2, header2)
+
+# Log parsing regexes
+class MultiqcModule(BaseMultiqcModule):
+
+    def __init__(self):
+
+        # Initialise the parent object
+        super(MultiqcModule, self).__init__(name='BISCUIT', anchor='biscuit',
+        href="https://github.com/zwdzwd/biscuit",
+        info="is a tool to map bisulfite converted sequence reads and determine"\
+        " cytosine methylation states.")
+
+        # Set up data structures
+        self.mdata = {
+            'mapq': {},
+            'markdup': {},
+            'coverage': {},
+            'retention': {},
+        }
+
+        # Find and parse bismark alignment reports
+        for k in self.mdata:
+            for f in self.find_log_files('biscuit/%s' % k):
+                d = getattr(self, 'parse_logs_%s' % k)(f['f'], f['fn'])
+                self.mdata[k][f['fn']] = d
+
+        num_parsed = 0
+        for k in self.mdata:
+            num_parsed += len(self.mdata[k])
+        if num_parsed == 0:
+            raise UserWarning
+
+        # Basic Stats Table
+        self.biscuit_stats_table()
+
+        # Write out to the report
+        for k in self.mdata:
+            if len(self.mdata[k]) > 0:
+                self.write_data_file(self.mdata[k], 'multiqc_biscuit_%s' % k)
+                log.info("Found %d biscuit %s reports" % (len(self.mdata[k]), k))
+                if k != 'markdup':
+                    getattr(self, 'chart_biscuit_%s' % k)()
+
+    def biscuit_stats_table(self):
+
+        pd = {}
+        for sid, dd in self.mdata['mapq'].items():
+            if not sid.endswith('_mapq_table.txt'):
```

Comment:
I'd prefer to remove this. Sometimes MultiQC users change the filenames of tool outputs and then customise their MultiQC config to update the search patterns. This will break things for them as the filename could not have that string in for genuine reasons.

If you want to validate that the correct thing has been parsed, I'd recommend doing this in the parsing step instead (preferably by checking file contents / parsed data).

Reply:
These are removed as mentioned above.